oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/ElasticsearchIndexDefinition.java

```diff
@@ -18,11 +18,100 @@
+
+    private static final int MAX_NAME_LENGTH = 255;
+
+    private static final String INVALID_CHARS_REGEX = Pattern.quote(INVALID_FILENAME_CHARS
+            .stream()
+            .map(Object::toString)
+            .collect(Collectors.joining("")));
+
+    private final String elasticsearchIndexName;
+
+    public final int bulkActions;
+    public final long bulkSizeBytes;
+    public final long bulkFlushIntervalMs;
+    public final int bulkRetries;
+    public final long bulkRetriesBackoff;
+
     public ElasticsearchIndexDefinition(NodeState root, NodeState defn, String indexPath) {
         super(root, getIndexDefinitionState(defn), determineIndexFormatVersion(defn), determineUniqueId(defn), indexPath);
+        this.elasticsearchIndexName = getRemoteIndexName();
+
+        this.bulkActions = getOptionalValue(defn, BULK_ACTIONS, BULK_ACTIONS_DEFAULT);
+        this.bulkSizeBytes = getOptionalValue(defn, BULK_SIZE_BYTES, BULK_SIZE_BYTES_DEFAULT);
+        this.bulkFlushIntervalMs = getOptionalValue(defn, BULK_FLUSH_INTERVAL_MS, BULK_FLUSH_INTERVAL_MS_DEFAULT);
+        this.bulkRetries = getOptionalValue(defn, BULK_RETRIES, BULK_RETRIES_DEFAULT);
+        this.bulkRetriesBackoff = getOptionalValue(defn, BULK_RETRIES_BACKOFF, BULK_RETRIES_BACKOFF_DEFAULT);
+    }
+
+    public String getElasticsearchIndexName() {
+        return elasticsearchIndexName;
+    }
+
+    private String getRemoteIndexName() {
+        // TODO: implement advanced remote index name strategy that takes into account multiple tenants and re-index process
+        return getESSafeIndexName(getIndexPath() + "-" + getReindexCount());
+    }
+
+    /**
+     * <ul>
+     *     <li>abc -> abc</li>
+     *     <li>xy:abc -> xyabc</li>
+     *     <li>/oak:index/abc -> abc</li>
+     * </ul>
+     * <p>
+     * The resulting file name would be truncated to MAX_NAME_LENGTH
+     */
+    private static String getESSafeIndexName(String indexPath) {
+        String name = StreamSupport
+                .stream(PathUtils.elements(indexPath).spliterator(), false)
+                .limit(3) //Max 3 nodeNames including oak:index which is the immediate parent for any indexPath
+                .filter(p -> !"oak:index".equals(p))
+                .map(ElasticsearchIndexDefinition::getESSafeName)
+                .collect(Collectors.joining("_"));
+
+        if (name.length() > MAX_NAME_LENGTH) {
+            name = name.substring(0, MAX_NAME_LENGTH);
+        }
+        return name;
+    }
+
+    /**
+     * Convert {@code e} to Elasticsearch safe index name.
+     * Ref: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html
+     */
+    private static String getESSafeName(String suggestedIndexName) {
+        return suggestedIndexName.replaceAll(INVALID_CHARS_REGEX, "").toLowerCase();
+    }
+
+    public static class Builder extends IndexDefinition.Builder {
```

Comment:
This is something we have inherited from the initial prototype that it's not completely clear to me. Despite the similarity in the names, these classes are needed for different reasons. I know the one in the `util` package is only needed for tests now but my understanding is that this could be useful when I have to create an elasticsearch index definition.

I have seen different ways to create an index definition. Not sure what is the more appropriate though.

@nit0906 @thomasmueller any advice on that? I would like to make it right now that we don't have many dependencies on this class.

Reply:
From how it's used, it looks like IndexDefinitionBuilder is used to create the index definition node structure (but not during indexing), and IndexDefinition.Builder is used to create an IndexDefinition object (but not the node structure). Adding javadocs would help!